### PR TITLE
[tests][dask] Add voting_parallel algorithm in tests (fixes #3834)

### LINKED ISSUE #3834: Support all LightGBM parallel tree learners in Dask
According to this piece of code 
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/python-package/lightgbm/dask.py#L251-L264
Dask module supports [all available tree learner types](https://lightgbm.readthedocs.io/en/latest/Parameters.html#tree_learner).
Refer to https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#choose-appropriate-parallel-algorithm for more details about difference among `tree_learner` param values.

However, our CI tests are run only for `data` parallel tree learner type.
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L182

I believe that parametrization of all tests with different tree learners will improve tests and make us confident in Dask module quality.

How tests can be parametrized:
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L31
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L166

### PR DESCRIPTION
**Pull request by jmoralez**
This includes the `voting_parallel` `tree_learner` for `test_regressor`, `test_classifier` and `test_ranker` in the tests for the dask module and removes the warning about experimental support that was previously triggered (because it wasn't tested).

### CODE REVIEW

File: tests/python_package_test/test_dask.py [hunk 1]
```diff
@@ -44,6 +44,7 @@
 CLIENT_CLOSE_TIMEOUT = 120
 
 tasks = ['binary-classification', 'multiclass-classification', 'regression', 'ranking']
+distributed_training_algorithms = ['data_parallel', 'voting_parallel']
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:47]: All 4 of these strings are valid for this argument (https://lightgbm.readthedocs.io/en/latest/Parameters.html#tree_learner):

* `data`
* `data_parallel`
* `voting`
* `voting_parallel`

But I don't think it's worth it to have all four choices here because that would create a lot of new tests (`2 * len(tasks) * len(data_output)`).

So could you do the following?

1. Change `distributed_training_algorithms` to the "main" values from https://lightgbm.readthedocs.io/en/latest/Parameters.html#tree_learner (`['data', 'voting_parallel']`)
2. add a new test like `test_training_respects_tree_learner_aliases()` that uses `['data_parallel', 'voting_parallel']`. It can just use `output="array"`, to reduce the total number of tests.

File: tests/python_package_test/test_dask.py [hunk 2]
```diff
@@ -905,22 +910,18 @@ def test_warns_and_continues_on_unrecognized_tree_learner(client):
...
+    X = da.random.random((1_000, 10), chunks=(500, 10))
+    y = da.random.random((1_000, 1), chunks=500)
+    dask_regressor = lgb.DaskLGBMRegressor(
+        client=client,
+        time_out=5,
+        tree_learner='feature_parallel',
+        n_estimators=1,
+        num_leaves=2
+    )
+    with pytest.raises(lgb.basic.LightGBMError):
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:923]:
```suggestion
    with pytest.raises(lgb.basic.LightGBMError, match="Do not support feature parallel in c api"):
```

Can you please add the specific error message in https://github.com/microsoft/LightGBM/issues/3834#issuecomment-782269920 here? In tests that catch errors, I think it's valuable to check that you got the exact error you expected. Otherwise, unexpected errors introduced by code changes can slip past the tests.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for doing this! I'm glad it was as easy as just changing the tests. I have a few suggested changes for the tests.

**Comment by jmoralez** [comment]
Hi, James. Do you have any suggestions on what to work on next?

**Comment by jmoralez** [comment]
I noticed a failing test due to graphviz I believe. This time it failed in [Linux_latest_gpu_source](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=9528&view=logs&j=582743ec-83fe-58c4-937e-4197b5816801&t=63b016c8-8965-547a-fab6-395af921c96a&l=628). Here are the errors:
```
Warning: Could not load "/home/AzDevOps_azpcontainer/miniconda/envs/test-env/lib/graphviz/libgvplugin_pango.so.6" - It was found, so perhaps one of its dependents was not.  Try ldd.
Warning: Could not load "/home/AzDevOps_azpcontainer/miniconda/envs/test-env/lib/graphviz/libgvplugin_gd.so.6" - It was found, so perhaps one of its dependents was not.  Try ldd.
```
I believe this test was also failing on some builds before I merged the latest master.

The other one is related to https://github.com/microsoft/LightGBM/pull/4095#issuecomment-804523725.

**Comment by jameslamb** [comment]
> Hi, James. Do you have any suggestions on what to work on next?

are you mostly interested in Dask? If so, this issue would be a good one to pick up next: https://github.com/microsoft/LightGBM/issues/3896. If not, let me know and I can recommend something else.

File: tests/python_package_test/test_dask.py (hunk 1 above)
Comment (jameslamb) [reply tests/python_package_test/test_dask.py]: looks like this was addressed in https://github.com/microsoft/LightGBM/pull/4088/commits/76b99941a8e8c016be9f375f09fe6d0af8e1108e

File: tests/python_package_test/test_dask.py (hunk 2 above)
Comment (jameslamb) [reply tests/python_package_test/test_dask.py]: looks like this was addressed in https://github.com/microsoft/LightGBM/pull/4088/commits/76b99941a8e8c016be9f375f09fe6d0af8e1108e

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for making that most recent round of changes. I don't have any other suggestions. Could you please update to the latest `master`?

File: tests/python_package_test/test_dask.py [hunk 3]
```diff
@@ -273,7 +276,7 @@ def test_classifier(output, task, boosting_type, client):
     p2_proba = local_classifier.predict_proba(X)
     s2 = local_classifier.score(X, y)
 
-    if boosting_type == 'rf' and output == 'dataframe-with-categorical':
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:276]: IS this change related to this PR?

File: tests/python_package_test/test_dask.py [hunk 4]
```diff
@@ -930,22 +935,36 @@ def test_warns_and_continues_on_unrecognized_tree_learner(client):
...
-        dask_regressor = lgb.DaskLGBMRegressor(
-            client=client,
-            time_out=5,
-            tree_learner=tree_learner,
-            n_estimators=1,
-            num_leaves=2
-        )
-        with pytest.warns(UserWarning, match='Support for tree_learner %s in lightgbm' % tree_learner):
-            dask_regressor = dask_regressor.fit(X, y)
+@pytest.mark.parametrize('task', tasks)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:938]: Can we not parametrize this test with `tasks` and not loose code coverage? It will be helpful to save some CI time.

File: tests/python_package_test/test_dask.py [hunk 5]
```diff
@@ -930,22 +935,36 @@ def test_warns_and_continues_on_unrecognized_tree_learner(client):
...
 
-        assert dask_regressor.fitted_
-        assert dask_regressor.get_params()['tree_learner'] == tree_learner
+    assert dask_model.fitted_
+    assert dask_model.get_params()['tree_learner'] == tree_learner
+
+
+def test_error_on_feature_parallel_tree_learner(client):
+    X = da.random.random((1_000, 10), chunks=(500, 10))
+    y = da.random.random((1_000, 1), chunks=500)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:958]: To save CI time.
```suggestion
    X = da.random.random((100, 10), chunks=(50, 10))
    y = da.random.random((100, 1), chunks=50)
```

**Review by StrikerRUS — COMMENTED** [review]
Great to know that voting parallel is working without any modifications!
Please consider checking a few my minor comments below:

File: tests/python_package_test/test_dask.py (hunk 3 above)
Comment (jmoralez) [reply tests/python_package_test/test_dask.py:276]: Yes. Because of https://github.com/microsoft/LightGBM/issues/4118, rf for classification isn't very reliable. When using `tree_learner='data'` these tests would pass, but with voting they all fail, so I moved all of the rf tests to the "soft check" instead of just the ones for `dataframe-with-categorical`.

File: tests/python_package_test/test_dask.py (hunk 4 above)
Comment (jmoralez) [reply tests/python_package_test/test_dask.py]: addressed in aa78ea3

File: tests/python_package_test/test_dask.py (hunk 5 above)
Comment (jmoralez) [reply tests/python_package_test/test_dask.py]: addressed in aa78ea3

**Comment by jmoralez** [comment]
@jameslamb do you know what's the advantage of using the `client` fixture? (Apart from avoiding the client construction). I just tried running `test_regressor` by initializing the client at the top and the time goes from 137s down to 98s.
Here's the profile with the fixture:
Screenshot from 2021-03-30 20-22-14
And without:
image
So basically with the fixture 70% of the time is spent in setting up and tearing down the cluster. Without it only 60% of the time is spent there and seems to be faster.

File: tests/python_package_test/test_dask.py (hunk 3 above)
Comment (StrikerRUS) [reply tests/python_package_test/test_dask.py:276]: Got it, thanks for the explanation!

**Comment by StrikerRUS** [comment]
@jmoralez 
> do you know what's the advantage of using the `client` fixture?

Awesome research! Could you please copy-paste your comment here: https://github.com/microsoft/LightGBM/issues/3829#issuecomment-769442538. I believe that place is the best one to continue the discussion.

**Review by StrikerRUS — APPROVED** [review]
LGTM!

**Review by jameslamb — APPROVED** [review]
thanks for this!
